torch_xla/distributed/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,157 @@
+    # Optimizer initialization
+    self.base_optimizer = optimizer_class(iter(self.sharded_params), **defaults)
+
+  def _shard_tensor(self, tensor: torch.Tensor):
+    """
+        Get the shard of the input tensor.
+        """
+    assert tensor.shape[0] % self.world_size == 0, "Not support padding now."
+    tensor = tensor.chunk(self.world_size)[self.rank]
+    return tensor
+
+  def _shard_parameters(self):
+    """
+        Shard all parameters.
+        """
+    xm.unlazy(self.params)
+    for param in self.params:
+      shard_data = param.data.to(device="cpu")  # move to cpu
+      shard_data = self._shard_tensor(shard_data)  # slice it
+      if shard_data.dtype != self.optimizer_dtype:
+        shard_data = shard_data.to(dtype=self.optimizer_dtype)
+      shard_data = shard_data.to(device=self.device)  # move to xla device
+      shard = nn.Parameter(shard_data, requires_grad=param.requires_grad)
+      self.sharded_params.append(shard)
+
+  @torch.no_grad()
+  def step(self, closure=None, **kwargs):
+    """
+        Performs a single optimizer step and syncs parameters across all ranks.
+        """
+    loss = None
+    if closure is not None:
+      with torch.enable_grad():
+        loss = closure()
+
+    # Reduce full gradients across ranks
+    # Assign gradient shards to the respective parameter shards
+    for param, shard in zip(self.params, self.sharded_params):
+      if param.grad is not None:
+        grad_shard = xm.reduce_scatter(
+            xm.REDUCE_SUM,
+            param.grad,
+            scale=1.0 / self.world_size,
+            scatter_dim=0,
+            shard_count=self.world_size,
+            pin_layout=self.pin_layout,
+            groups=self.cc_op_groups,
+        )
+
+        if grad_shard.dtype != self.optimizer_dtype:
+          grad_shard = grad_shard.to(dtype=self.optimizer_dtype)
+        shard.grad = grad_shard
+
+    if self.grad_clipping:
+      # Update unscale/clip with sub partitions
+      torch.nn.utils.clip_grad_norm_(
+          self.sharded_params, max_norm=self.max_norm)
+
+    # Step the wrapped optimizer
+    loss = self.base_optimizer.step(closure=closure, **kwargs)
+    # Remove shards' grads
+    self.base_optimizer.zero_grad(set_to_none=True)
+
+    # All gather the new weights across the ranks and assign them to the full parameters
+    for param, shard in zip(self.params, self.sharded_params):
+      if param.grad is not None:
+        shard_data = shard.data
+        if param.dtype != self.optimizer_dtype:
+          shard_data = shard_data.to(dtype=param.dtype)
+        xm.all_gather(
+            shard_data,
+            dim=0,
+            output=param.data,
+            pin_layout=self.pin_layout,
+            groups=self.cc_op_groups,
+        )
+
+    return loss
+
+  def state_dict(self):
+    state_dict = super().state_dict()
+    state_dict['base'] = self.base_optimizer.state_dict()
+    return state_dict
+
+  def load_state_dict(self, state_dict):
```

Comment:
do we need test coverage for `state_dict` and `load_state_dict`?

Reply:
yes, i'll add them